src/librustc_parse/parser/item.rs

```diff
@@ -1739,7 +1739,8 @@ impl<'a> Parser<'a> {
     fn recover_nested_adt_item(&mut self, keyword: Symbol) -> PResult<'a, bool> {
         if self.token.is_keyword(kw::Enum) ||
             self.token.is_keyword(kw::Struct) ||
-            self.token.is_keyword(kw::Union)
+            self.token.is_keyword(kw::Union) &&
+            self.next_tok().is_ident()
```

Comment:
```suggestion
            self.look_ahead(1, |t| t.is_ident())
```
`next_tok` is a more low-level operation; prefer the higher level operation for looking-ahead :)

Reply:
done